File: scripts/import_produtos.py
#!/usr/bin/env python3

import re
import os
import sys
import json

# Ler arquivo de produtos
with open('client/src/data/produtos.ts', 'r', encoding='utf-8') as f:
    content = f.read()

# Extrair array de produtos
match = re.search(r'export const produtos: Produto\[\] = \[([\s\S]*?)\];', content)
if not match:
    print('❌ Não foi possível encontrar o array de produtos')
    sys.exit(1)

array_content = match.group(1)

# Remover comentários
array_content = re.sub(r'//.*$', '', array_content, flags=re.MULTILINE)

# Converter para JSON válido
# 1. Adicionar aspas em torno das chaves
array_content = re.sub(r'(\w+):', r'"\1":', array_content)

# 2. Remover vírgulas antes de } ou ]
array_content = re.sub(r',\s*([}\]])', r'\1', array_content)

# 3. Remover quebras de linha extras
array_content = re.sub(r'\n\s*\n', '\n', array_content)

# 4. Remover espaços em branco extras
array_content = re.sub(r'\s+', ' ', array_content)

# Fazer parse do JSON
try:
    json_str = '[' + array_content + ']'
    produtos = json.loads(json_str)
    print(f'✅ Encontrados {len(produtos)} produtos')
except json.JSONDecodeError as e:
    print(f'❌ Erro ao fazer parse do JSON: {e}')
    # Tentar um parse mais permissivo
    print('Tentando parse alternativo...')
    
    # Extrair objetos individuais
    produtos = []
    # Procurar por padrão {id: X, ... }
    for obj_match in re.finditer(r'\{[^{}]*id:\s*\d+[^{}]*\}', array_content):
        obj_str = obj_match.group(0)
        obj_str = re.sub(r'(\w+):', r'"\1":', obj_str)
        try:
            obj = json.loads(obj_str)
            produtos.append(obj)
        except:
            pass
    
    if not produtos:
        print('❌ Não foi possível fazer parse dos produtos')
        sys.exit(1)
    
    print(f'✅ Encontrados {len(produtos)} produtos (parse alternativo)')

# Gerar SQL
sql_lines = ['DELETE FROM produtos;']

for p in produtos:
    try:
        codigo = str(p.get('codigo', '')).replace("'", "''")
        descricao = str(p.get('descricao', '')).replace("'", "''")
        categoria = str(p.get('categoria', '')).replace("'", "''")
        ncm = str(p.get('ncm', '')).replace("'", "''")
        cod_barras = str(p.get('cod_barras', '')).replace("'", "''")
        unid = str(p.get('unid', 'UND')).replace("'", "''")
        caixa = str(p.get('caixa', 'PAR')).replace("'", "''")
        volt = str(p.get('volt', 'BIVOLT')).replace("'", "''")
        custo_usd = float(p.get('custo_usd', 0))
        preco_venda = float(p.get('preco_venda', 0))
        
        sql = f"""INSERT INTO produtos (codigo, descricao, categoria, unidade, caixa, voltagem, codigoBarras, ncm, custoUsd, precoVendaBrl, ativo) 
VALUES ('{codigo}', '{descricao}', '{categoria}', '{unid}', '{caixa}', '{volt}', '{cod_barras}', '{ncm}', {custo_usd}, {preco_venda}, 'true');"""
        
        sql_lines.append(sql)
    except Exception as e:
        print(f'⚠️  Erro ao processar produto {p.get("codigo")}: {e}')
        continue

# Salvar SQL
with open('/tmp/import_produtos.sql', 'w', encoding='utf-8') as f:
    f.write('\n'.join(sql_lines))

print(f'✅ Arquivo SQL criado com {len(sql_lines)-1} INSERT statements')
print(f'📁 Arquivo: /tmp/import_produtos.sql')
print(f'📊 Tamanho: {os.path.getsize("/tmp/import_produtos.sql")} bytes')

File: scripts/import-produtos.mjs
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import mysql from 'mysql2/promise';
import dotenv from 'dotenv';

// Carregar variáveis de ambiente
dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Função para converter TypeScript para JSON válido
function convertTsToJson(tsContent) {
  // Remover comentários de linha
  let json = tsContent.replace(/\/\/.*$/gm, '');
  
  // Remover quebras de linha extras
  json = json.replace(/\n\s*\n/g, '\n');
  
  // Converter propriedades sem aspas para com aspas
  // Ex: id: 1 -> "id": 1
  json = json.replace(/(\w+):/g, '"$1":');
  
  // Remover vírgulas antes de } ou ]
  json = json.replace(/,\s*([}\]])/g, '$1');
  
  return json;
}

async function importProdutos() {
  let connection;
  try {
    // Ler arquivo de produtos
    const produtosPath = path.join(__dirname, '../client/src/data/produtos.ts');
    let produtosContent = fs.readFileSync(produtosPath, 'utf-8');

    // Extrair apenas o array de produtos
    let arrayStart = produtosContent.indexOf('export const produtos: Produto[] = [');
    let arrayEnd = produtosContent.lastIndexOf('];');
    
    if (arrayStart === -1 || arrayEnd === -1) {
      console.error('❌ Não foi possível encontrar o array de produtos');
      process.exit(1);
    }

    // Extrair o conteúdo do array
    let arrayContent = produtosContent.substring(arrayStart + 'export const produtos: Produto[] = '.length, arrayEnd + 1);

    // Converter TypeScript para JSON válido
    let jsonContent = convertTsToJson(arrayContent);

    // Fazer parse do JSON
    let produtos = [];
    try {
      produtos = JSON.parse(jsonContent);
    } catch (e) {
      console.error('❌ Erro ao fazer parse do JSON:', e.message);
      // Mostrar um trecho do JSON para debug
      console.error('Conteúdo convertido (primeiros 500 chars):', jsonContent.substring(0, 500));
      process.exit(1);
    }

    console.log(`✅ Encontrados ${produtos.length} produtos para importar`);

    // Conectar ao banco de dados
    connection = await mysql.createConnection({
      host: process.env.DB_HOST || 'localhost',
      user: process.env.DB_USER || 'root',
      password: process.env.DB_PASSWORD || '',
      database: process.env.DB_NAME,
      port: process.env.DB_PORT || 3306,
    });

    console.log('✅ Conectado ao banco de dados');

    // Limpar tabela de produtos existentes
    await connection.execute('DELETE FROM produtos');
    console.log('🗑️  Tabela de produtos limpa');

    // Preparar insert
    let importedCount = 0;
    let errorCount = 0;

    for (const produto of produtos) {
      try {
        await connection.execute(
          `INSERT INTO produtos (
            codigo, descricao, categoria, unidade, caixa, voltagem,
            codigoBarras, ncm, custoUsd, precoVendaBrl, ativo
          ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
          [
            produto.codigo,
            produto.descricao,
            produto.categoria,
            produto.unid || 'UND',
            produto.caixa || 'PAR',
            produto.volt || 'BIVOLT',
            produto.cod_barras || null,
            produto.ncm || null,
            parseFloat(produto.custo_usd) || 0,
            parseFloat(produto.preco_venda) || 0,
            'true',
          ]
        );
        importedCount++;
      } catch (error) {
        console.error(`❌ Erro ao importar produto ${produto.codigo}:`, error.message);
        errorCount++;
      }
    }

    console.log(`\n📊 Resultado da importação:`);
    console.log(`✅ Importados: ${importedCount} produtos`);
    console.log(`❌ Erros: ${errorCount} produtos`);

    // Verificar contagem final
    const [rows] = await connection.execute('SELECT COUNT(*) as count FROM produtos');
    console.log(`📈 Total de produtos no banco: ${rows[0].count}`);

    await connection.end();
    console.log('\n✅ Importação concluída com sucesso!');
  } catch (error) {
    console.error('❌ Erro durante importação:', error);
    process.exit(1);
  }
}

// Executar importação
importProdutos();
